### PR TITLE
rafthttp: use http.Request.WithContext instead of Cancel

### PR DESCRIPTION
**Pull request by sinsharat**
#6953

### CODE REVIEW

**Comment by sinsharat** [comment]
@xiang90 @heyitsanthony PTAL. Don't think the jenkins-proxy-ci is failing due to this change. Have already tested e2e, rafthttptest and integration in my local machine.
Thanks!

File: rafthttp/pipeline.go [hunk 1]
```diff
@@ -22,11 +22,12 @@ import (
 	"time"
 
 	"github.com/coreos/etcd/etcdserver/stats"
-	"github.com/coreos/etcd/pkg/httputil"
 	"github.com/coreos/etcd/pkg/pbutil"
 	"github.com/coreos/etcd/pkg/types"
 	"github.com/coreos/etcd/raft"
 	"github.com/coreos/etcd/raft/raftpb"
+
+	"golang.org/x/net/context"
```
Comment (xiang90) [inline rafthttp/pipeline.go]: use https://golang.org/pkg/context/ instead?

File: rafthttp/pipeline.go [hunk 2]
```diff
@@ -128,31 +132,43 @@ func (p *pipeline) post(data []byte) (err error) {
...
-	resp.Body.Close()
-
-	err = checkPostResponse(resp, b, req, p.peerID)
-	if err != nil {
-		p.picker.unreachable(u)
-		// errMemberRemoved is a critical error since a removed member should
-		// always be stopped. So we use reportCriticalError to report it to errorc.
-		if err == errMemberRemoved {
-			reportCriticalError(err, p.errorc)
+	go func() {
```
Comment (xiang90) [inline rafthttp/pipeline.go]: why put this into a go routine? what's wrong with the previous code?

File: rafthttp/snapshot_sender.go [hunk 3]
```diff
@@ -26,6 +26,8 @@ import (
 	"github.com/coreos/etcd/pkg/types"
 	"github.com/coreos/etcd/raft"
 	"github.com/coreos/etcd/snap"
+
+	"golang.org/x/net/context"
```
Comment (xiang90) [inline rafthttp/snapshot_sender.go]: use https://golang.org/pkg/context/

File: rafthttp/stream.go [hunk 4]
```diff
@@ -410,6 +412,7 @@ func (cr *streamReader) stop() {
 }
 
 func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
+
```
Comment (xiang90) [inline rafthttp/stream.go]: remove this extra empty line

**Comment by xiang90** [comment]
Ignore proxy-ci for now.

**Comment by heyitsanthony** [comment]
@sinsharat looks like a new failure. I tried to reproduce with `go test -run=RejectInsecure -v -tags cluster_proxy` from your brach but it seems OK...

File: rafthttp/pipeline.go (hunk 1 above)
Comment (sinsharat) [inline rafthttp/pipeline.go]: sure

File: rafthttp/pipeline.go (hunk 2 above)
Comment (sinsharat) [inline rafthttp/pipeline.go]: @xiang90 i have coded a select statement after this so that after cancel the program exits. Without this the test case was failing on using request.WithCancel.

File: rafthttp/snapshot_sender.go (hunk 3 above)
Comment (sinsharat) [inline rafthttp/snapshot_sender.go]: sure.

File: rafthttp/stream.go (hunk 4 above)
Comment (sinsharat) [inline rafthttp/stream.go]: sure

File: rafthttp/pipeline.go (hunk 2 above)
Comment (sinsharat) [inline rafthttp/pipeline.go]: @xiang90 i tested it with old code, it will not work. The test case fails. So this code is required for it to work properly in this case.
image

**Comment by sinsharat** [comment]
@xiang90 @heyitsanthony all changes done. PTAL. 
Thanks!

File: rafthttp/pipeline.go (hunk 2 above)
Comment (xiang90) [inline rafthttp/pipeline.go]: Why the old work works? But now it does not? I think we should figure it out. The only change how to set the cancel(), right? It should not affect the code after it I assume.

Comment (sinsharat) [inline rafthttp/pipeline.go]: @xiang90 I had already checked the code earlier of go package. With old code it checks for cancel. But with the new implementation it needs the check for ctx.Done. The above code ensures the proper closure of request before exiting. Without this code, it doesn't.

Comment (xiang90) [inline rafthttp/pipeline.go]: I do not think I understand what you said.

Let me clarify it.

now, we:

- create context with cancel at line 122
- pass in the created context to req at line 123

Now if the cancel fun get called, the request should be canceled.

- create a go routine to listen on stop. if server is stopped, we can cancel.

So if server is stopped, go routine should be triggered on stop channel, and call cancel.

After it calls cancel, the http request should be canceled.

From your test, it fails to cancel the http request. There is something goes wrong.

You "fix" that by making the http request nonblocking. That is not what we want. We want the request to be canceled once the server is stopped. Basically we need to make sure the go routine created at 134 exits.

Comment (sinsharat) [inline rafthttp/pipeline.go]: @xiang90 ok will check it out.

Comment (sinsharat) [inline rafthttp/pipeline.go]: @xiang90 i went through the code and i found that the issue is with the test case. Earlier the request cancel was being handled by us manually, hence the roundTripperBlocker used for test case is checking for req.Cancel which used to come earlier from our manual implementation. Now that is not the case and since the roundTripperBlocker has no implementation for ctx.Done hence it was failing. 
Will correct the same and update the PR. Thanks!

**Comment by sinsharat** [comment]
@xiang90 @heyitsanthony All fixed. PTAL.
Thanks!

File: rafthttp/snapshot_sender.go [hunk 5]
```diff
@@ -104,7 +105,9 @@ func (s *snapshotSender) send(merged snap.Message) {
 // post posts the given request.
 // It returns nil when request is sent out and processed successfully.
 func (s *snapshotSender) post(req *http.Request) (err error) {
-	cancel := httputil.RequestCanceler(req)
+	ctx, cancel := context.WithCancel(context.Background())
+	req = req.WithContext(ctx)
+	defer cancel()
```
Comment (xiang90) [inline rafthttp/snapshot_sender.go]: revert this change ( defer cancel() )? or change L128 of rafthttp/pipeline.go to make them consistent?

**Comment by xiang90** [comment]
looks great. there are a few minor issues. we need to continue to clean up other pkgs if there is any. Thanks @sinsharat!

File: rafthttp/snapshot_sender.go (hunk 5 above)
Comment (sinsharat) [inline rafthttp/snapshot_sender.go]: @xiang90 i had to do this change because the govet in CI was giving a possible go routine leak error with old cancel placed there since it expects cancel to be called mandatorily before exiting. What do you want me to do for this?

**Comment by sinsharat** [comment]
@xiang90 do you want other packages to be corrected in the same PR?

**Comment by xiang90** [comment]
@sinsharat No. We can do them in other prs :P

**Comment by sinsharat** [comment]
@xiang90 Let me know if anything else required to be done here.
Thank you!

File: rafthttp/snapshot_sender.go (hunk 5 above)
Comment (xiang90) [inline rafthttp/snapshot_sender.go]: can you also fix other places? like L128 of rafthttp/pipeline.go

Comment (sinsharat) [inline rafthttp/snapshot_sender.go]: @xiang90 for that we will need to change the design a bit since we cannot use defer directly there. For this reason govmt is not giving the error there. Using defer will not cause the code to work properly. So we will need to move the go routine code to i.e. L124 to L128 outside go routine and the code from L132 to L156 inside the go routine. Is that fine with you?

Comment (xiang90) [inline rafthttp/snapshot_sender.go]: ok. we can deal with that later.

**Comment by xiang90** [comment]
LGTM

**Comment by xiang90** [comment]
@sinsharat Actually can you squash two commits into one?

File: rafthttp/snapshot_sender.go (hunk 5 above)
Comment (sinsharat) [inline rafthttp/snapshot_sender.go]: ok

**Comment by sinsharat** [comment]
@xiang90 its squashed into one.
Thanks!

**Comment by xiang90** [comment]
Thank you!

File: rafthttp/pipeline.go [hunk 6]
```diff
@@ -16,13 +16,13 @@ package rafthttp
 
 import (
 	"bytes"
+	"context"
```
Comment (fanminshi) [inline rafthttp/pipeline.go]: should we import `"golang.org/x/net/context"` instead of `context`? @xiang90 @heyitsanthony .

Comment (xiang90) [inline rafthttp/pipeline.go]: @fanminshi we will use context in 3.2 release. we do not need the compatibility anymore.

Comment (fanminshi) [inline rafthttp/pipeline.go]: got it. Also have we refractor our code to use `context` yet?
